pkg/chunk/cache/fifo_cache.go

```diff
@@ -166,108 +190,93 @@ func (c *FifoCache) Store(ctx context.Context, keys []string, bufs [][]byte) {
 
 // Stop implements Cache.
 func (c *FifoCache) Stop() {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	c.entriesEvicted.Add(float64(c.lru.Len()))
+
+	c.entries = make(map[string]*list.Element)
+	c.lru.Init()
+	c.currSizeBytes = 0
+
+	c.entriesCurrent.Set(float64(0))
+	c.memoryBytes.Set(float64(0))
 }
 
 // Put stores the value against the key.
 func (c *FifoCache) Put(ctx context.Context, keys []string, values []interface{}) {
 	c.entriesAdded.Inc()
-	if c.size == 0 {
-		return
-	}
 
 	c.lock.Lock()
 	defer c.lock.Unlock()
 
 	for i := range keys {
-		c.put(ctx, keys[i], values[i])
+		c.put(keys[i], values[i])
 	}
 }
 
-func (c *FifoCache) put(ctx context.Context, key string, value interface{}) {
-	// See if we already have the entry
-	index, ok := c.index[key]
+func (c *FifoCache) put(key string, value interface{}) {
+	// See if we already have the item in the cache.
+	element, ok := c.entries[key]
 	if ok {
-		entry := c.entries[index]
-		deltaSize := sizeOf(value) - sizeOf(entry.value)
-
-		entry.updated = time.Now()
-		entry.value = value
+		// Remove the item from the cache.
+		entry := c.lru.Remove(element).(*cacheEntry)
+		delete(c.entries, key)
+		c.currSizeBytes -= sizeOf(entry)
```

Comment:
I think we need to cache entry size from moment when entry is inserted to cache. Entry size may have changed since then, and then this computation will be wrong.

Reply:
Not sure I understand. We have three types of items: the old cache item for the existing key, the evicted item(s), and the new cache item for the existing key. All three types are tracked correctly IMO. The `c.currSizeBytes` is updated the moment the cache content is changed.